src/psyclone/domain/nemo/transformations/nemo_arrayaccess2loop_trans.py

```diff
@@ -0,0 +1,328 @@
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node, but found '{0}'.".format(type(node.parent).__name__))
+        array_ref = node.parent
+        # Array reference not within an assignment
+        if not array_ref.parent or not isinstance(array_ref.parent,
+                                                  Assignment):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within an Assignment node, but found '{0}'."
+                .format(type(array_ref.parent).__name__))
+        assignment = array_ref.parent
+        # Array reference not on lhs of the assignment
+        if assignment.lhs is not array_ref:
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within the left-hand-side of an Assignment "
+                "node, but it is on the right-hand-side.")
+
+        # Contains a range node
+        if node.walk(Range):
+            raise TransformationError(
+                "Error in NemoArrayAccess2LoopTrans transformation. The "
+                "supplied node should not be or contain a Range node as it "
+                "should be single valued.")
+
+        # Capture loop iterator symbols in order
+        iterator_symbols = []
+        location = node.parent.parent
+        while (isinstance(location.parent, Schedule) and
+               isinstance(location.parent.parent, (Loop, NemoKern))):
+            location = location.parent.parent
+            if isinstance(location, Loop):
+                iterator_symbols.append(location.variable)
+
+        # Index contains a loop iterator
+        for reference in node.walk(Reference):
+            if reference.symbol in iterator_symbols:
+                raise TransformationError(
+                    "Error in NemoArrayAccess2LoopTrans transformation. The "
+                    "supplied node should not be or contain a loop iterator, "
+                    "it should be single valued.")
+
+        # An index before this one has more than one loop iterator
+        array_access_iterators = []
+        for index in node.parent.children[:node.position]:
+            # Find any iterators in this index
+            index_iterators = [
+                my_index.symbol for my_index in index.walk(Reference)
+                if my_index.symbol in iterator_symbols]
+            iterator_names = [my_index.name for my_index in index_iterators]
+            if len(index_iterators) > 1:
+                raise TransformationError(
```

Comment:
I don't understand this restriction. Is it because we're trying to insert the new loop somewhere higher up the tree? If so, I think it's another reason not to attempt to be clever!

Reply:
Yes, this and the next two tests were constraints that check that the code conforms to the expected looping structure which is required when placing a new loop in the appropriate place. I've removed these as constraints and added them as tests that work. There is one case that needs to be checked though. The NEMO API provides an expected loop iterator name. This name should not be used for iterating over any other indices. I've added a validation check and test for this. i.e. `do jj=1,n;a(jj,1);end do` will raise an exception as `jj` should be used for the second loop index.